src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,125 @@ public SubstitutionMatrix(final byte[] matrix) {
+            }
+            stringBuilder.append('\t');
+        }
+        // lower case substitutions
+        for (final SubstitutionBase r : BASES) {
+            char lowerCaseBase = Character.toLowerCase((char)r.getBase());
+            stringBuilder.append(lowerCaseBase);
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[lowerCaseBase][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        return stringBuilder.toString();
+    }
+
+    // populate a matrix of substitution frequencies from a list of CramCompressionRecords with Substitution features
+    private static long[][] buildFrequencies(final List<CramCompressionRecord> cramRecords) {
+        final long[][] frequencies = new long[SYMBOL_SPACE_SIZE][SYMBOL_SPACE_SIZE];
+        for (final CramCompressionRecord record : cramRecords) {
+            if (record.readFeatures != null) {
+                for (final ReadFeature readFeature : record.readFeatures) {
+                    if (readFeature.getOperator() == Substitution.operator) {
+                        final Substitution substitution = ((Substitution) readFeature);
+                        final byte refBase = substitution.getReferenceBase();
+                        final byte base = substitution.getBase();
+                        if (refBase <= 0 || base <= 0) {
+                            throw new IllegalArgumentException(
+                                    String.format("CRAM: Attempt to generate a substitution code for invalid reference base '%c'", (char) refBase));
+                        }
+                        frequencies[refBase][base]++;
+                    }
+                }
+            }
+        }
+        return frequencies;
+    }
+
     private static class SubstitutionFrequency {
-        final byte base;
+        final SubstitutionBase substituteBase;
         long freq;
         byte rank;
 
-        public SubstitutionFrequency(final byte base, final long freq) {
-            this.base = base;
+        public SubstitutionFrequency(final SubstitutionBase substituteBase, final long freq) {
+            this.substituteBase = substituteBase;
             this.freq = freq;
         }
     }
 
-    private static final Comparator<SubstitutionFrequency> comparator = new Comparator<SubstitutionFrequency>() {
-
-        @Override
-        public int compare(final SubstitutionFrequency o1, final SubstitutionFrequency o2) {
-            // primary sort by frequency
-            if (o1.freq != o2.freq) {
-                return (int) (o2.freq - o1.freq);
-            }
-            // same frequency; compare based on spec tie-breaking rule (use base order prescribed by the spec)
-            return CODE_ORDER[o1.base] - CODE_ORDER[o2.base];
-        }
-    };
+    private static final Comparator<SubstitutionFrequency> comparator =
```

Comment:
Use uppercase for statics
```suggestion
    private static final Comparator<SubstitutionFrequency> COMPARATOR =
```

Reply:
Done.